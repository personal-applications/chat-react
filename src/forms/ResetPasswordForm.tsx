import { ajvResolver } from "@hookform/resolvers/ajv";
import { JSONSchemaType } from "ajv";
import { fullFormats } from "ajv-formats/dist/formats";
import { useForm } from "react-hook-form";
import { If, Then } from "react-if";
import { Link, useSearchParams } from "react-router-dom";
import PasswordInput from "../components/PasswordInput";
import { PASSWORD_REGEX } from "../constants";
import { ServerError, isServerFromError } from "../services/error";

export type ResetPasswordFields = {
  newPassword: string;
  confirmPassword: string;
  token?: string;
};

const schema: JSONSchemaType<ResetPasswordFields> = {
  type: "object",
  properties: {
    newPassword: { type: "string", pattern: PASSWORD_REGEX },
    confirmPassword: { type: "string", pattern: PASSWORD_REGEX },
    token: { type: "string", nullable: true },
  },
  required: ["newPassword", "confirmPassword"],
  additionalProperties: false,
  errorMessage: {
    properties: {
      password:
        "Password must be at least 8 characters long and contain at least one uppercase letter, one lowercase letter, one number, and one special character.",
      confirmPassword: "Passwords do not match.",
    },
  },
};

export type Prop = {
  onSubmit: (data: ResetPasswordFields) => Promise<{ message: string }>;
};

export default function ResetPasswordForm({ onSubmit }: Prop) {
  const [searchParams] = useSearchParams();
  const {
    register,
    handleSubmit,
    formState: { errors, isSubmitting },
    setError,
  } = useForm<ResetPasswordFields>({
    resolver: ajvResolver(schema, {
      formats: fullFormats,
      removeAdditional: true,
    }),
    reValidateMode: "onChange",
  });

  const asyncOnSubmit = async (data: ResetPasswordFields) => {
    try {
      data.token = searchParams.get("token") ?? undefined;
      const { message } = await onSubmit(data);
      setError("root.serverResponse", { message });
    } catch (error) {
      if (isServerFromError(error)) {
        const e = error as ServerError;
        setError("root.serverError", {
          message: e.message,
        });

        return;
      }

      setError("root.serverError", {
        message: "An unknown error occurred.",
      });
    }
  };

  return (
    <form
      noValidate
      onSubmit={handleSubmit(asyncOnSubmit)}
      className="mt-6 flex flex-col gap-y-6 rounded-md bg-base-100 p-8"
    >
      <If condition={!isSubmitting && !!errors?.root?.serverResponse}>
        <Then>
          <div role="alert" className="alert alert-success">
            <svg
              xmlns="http://www.w3.org/2000/svg"
              className="stroke-current shrink-0 h-6 w-6"
              fill="none"
              viewBox="0 0 24 24"
            >
              <path
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                d="M9 12l2 2 4-4m6 2a9 9 0 11-18 0 9 9 0 0118 0z"
              />
            </svg>
            <span>{errors.root?.serverResponse?.message}</span>
          </div>
        </Then>
      </If>
      <If condition={!!errors?.root?.serverError}>
        <Then>
          <div role="alert" className="alert alert-error">
            <svg
              xmlns="http://www.w3.org/2000/svg"
              className="stroke-current shrink-0 h-6 w-6"
              fill="none"
              viewBox="0 0 24 24"
            >
              <path
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                d="M10 14l2-2m0 0l2-2m-2 2l-2-2m2 2l2 2m7-2a9 9 0 11-18 0 9 9 0 0118 0z"
              />
            </svg>
            <span>{errors.root?.serverError?.message}</span>
          </div>
        </Then>
      </If>

      <PasswordInput
        placeholder="Password"
        error={errors.newPassword?.message}
        {...register("newPassword")}
      />

      <PasswordInput
        placeholder="Confirm Password"
        error={errors.confirmPassword?.message}
        {...register("confirmPassword")}
      />

      <Link to="/forgot-password" className="link link-primary link-hover">
        Forgot password?
      </Link>
      <button className="btn btn-primary" type="submit" disabled={isSubmitting}>
        Reset
      </button>
    </form>
  );
}
