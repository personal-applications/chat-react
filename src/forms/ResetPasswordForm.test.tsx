import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { BrowserRouter } from "react-router-dom";
import ResetPasswordForm, { ResetPasswordFields } from "./ResetPasswordForm";

const mockOnSubmit = vi.fn();

const renderForm = () => {
  render(
    <BrowserRouter>
      <ResetPasswordForm onSubmit={mockOnSubmit} />
    </BrowserRouter>,
  );
};

test("renders form with password inputs and submit button", () => {
  renderForm();

  const passwordInput = screen.getByPlaceholderText("New Password");
  const confirmPasswordInput = screen.getByPlaceholderText("Confirm Password");
  const submitButton = screen.getByRole("button", { name: "Reset" });

  expect(passwordInput).toBeInTheDocument();
  expect(confirmPasswordInput).toBeInTheDocument();
  expect(submitButton).toBeInTheDocument();
});

test("displays error message when passwords do not match", async () => {
  renderForm();

  const passwordInput = screen.getByPlaceholderText("New Password");
  const confirmPasswordInput = screen.getByPlaceholderText("Confirm Password");
  const submitButton = screen.getByRole("button", { name: "Reset" });

  // Enter different passwords
  await userEvent.type(passwordInput, "password1");
  await userEvent.type(confirmPasswordInput, "password2");

  // Submit the form
  await userEvent.click(submitButton);

  const errorMessage = screen.getByText("Passwords do not match.");
  expect(errorMessage).toBeInTheDocument();

  expect(mockOnSubmit).not.toHaveBeenCalled();
});

test("calls onSubmit with form data when passwords match", async () => {
  renderForm();

  const passwordInput = screen.getByPlaceholderText("New Password");
  const confirmPasswordInput = screen.getByPlaceholderText("Confirm Password");
  const submitButton = screen.getByRole("button", { name: "Reset" });

  // Enter different passwords
  await userEvent.type(passwordInput, "Aa123456*");
  await userEvent.type(confirmPasswordInput, "Aa123456*");

  // Submit the form
  await userEvent.click(submitButton);

  expect(mockOnSubmit).toHaveBeenCalledTimes(1);
  expect(mockOnSubmit).toHaveBeenCalledWith({
    newPassword: "Aa123456*",
    confirmPassword: "Aa123456*",
  } as ResetPasswordFields);
});

test("displays success message when form submission is successful", async () => {
  mockOnSubmit.mockResolvedValueOnce({ message: "Password reset successful." });

  renderForm();

  const passwordInput = screen.getByPlaceholderText("New Password");
  const confirmPasswordInput = screen.getByPlaceholderText("Confirm Password");
  const submitButton = screen.getByRole("button", { name: "Reset" });

  // Enter different passwords
  await userEvent.type(passwordInput, "Aa123456*");
  await userEvent.type(confirmPasswordInput, "Aa123456*");

  // Submit the form
  await userEvent.click(submitButton);

  const successMessage = screen.getByText("Password reset successful.");
  expect(successMessage).toBeInTheDocument();
});

test("displays error message when form submission fails", async () => {
  mockOnSubmit.mockRejectedValueOnce(new Error("An unknown error occurred."));

  renderForm();

  const passwordInput = screen.getByPlaceholderText("New Password");
  const confirmPasswordInput = screen.getByPlaceholderText("Confirm Password");
  const submitButton = screen.getByRole("button", { name: "Reset" });

  // Enter different passwords
  await userEvent.type(passwordInput, "Aa123456*");
  await userEvent.type(confirmPasswordInput, "Aa123456*");

  // Submit the form
  await userEvent.click(submitButton);

  const errorMessage = screen.getByText("An unknown error occurred.");
  expect(errorMessage).toBeInTheDocument();
});

test("renders ResetPasswordForm correctly", () => {
  const tree = renderForm();

  expect(tree).toMatchSnapshot();
});
