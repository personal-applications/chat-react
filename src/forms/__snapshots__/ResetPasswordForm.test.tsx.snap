// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders ResetPasswordForm correctly 1`] = `undefined`;
